{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data exploration and visualization for scientific paper\n",
    "Analysis of the output files generated by Colocator for rheumatoid arthritis (RA) and type 1 diabetes (T1D).\n",
    "For plotting there can't duplicate names for the gene_names. If they are, change the name in the original file to the ensmbl name for example."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1002\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  var JS_MIME_TYPE = 'application/javascript';\n  var HTML_MIME_TYPE = 'text/html';\n  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  var CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    var script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    var cell = handle.cell;\n\n    var id = cell.output_area._bokeh_element_id;\n    var server_id = cell.output_area._bokeh_server_id;\n    // Clean up Bokeh references\n    if (id != null && id in Bokeh.index) {\n      Bokeh.index[id].model.document.clear();\n      delete Bokeh.index[id];\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd, {\n        iopub: {\n          output: function(msg) {\n            var id = msg.content.text.trim();\n            if (id in Bokeh.index) {\n              Bokeh.index[id].model.document.clear();\n              delete Bokeh.index[id];\n            }\n          }\n        }\n      });\n      // Destroy server and session\n      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    var output_area = handle.output_area;\n    var output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      var bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      var script_attrs = bk_div.children[0].attributes;\n      for (var i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      var toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    var events = require('base/js/events');\n    var OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1002\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    const hashes = {\"https://cdn.bokeh.org/bokeh/release/bokeh-2.3.2.min.js\": \"XypntL49z55iwGVUW4qsEu83zKL3XEcz0MjuGOQ9SlaaQ68X/g+k1FcioZi7oQAc\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.3.2.min.js\": \"bEsM86IHGDTLCS0Zod8a8WM6Y4+lafAL/eSiyQcuPzinmWNgNO2/olUF0Z2Dkn5i\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.3.2.min.js\": \"TX0gSQTdXTTeScqxj6PVQxTiRW8DOoGVwinyi1D3kxv7wuxQ02XkOxv0xwiypcAH\"};\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      if (url in hashes) {\n        element.crossOrigin = \"anonymous\";\n        element.integrity = \"sha384-\" + hashes[url];\n      }\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  var js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.3.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.3.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.3.2.min.js\"];\n  var css_urls = [];\n  \n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (var i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from bokeh.plotting import figure, show\n",
    "from bokeh.io import output_notebook\n",
    "output_notebook()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the data\n",
    "Fill here in the name of your file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "df_reuma = pd.read_csv('ra_output.csv')\n",
    "df_t1d = pd.read_csv('t1d_output.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data exploration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>tested_snps</th>\n",
       "      <th>pph4</th>\n",
       "      <th>top_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TRIM26</td>\n",
       "      <td>ENSG00000234127</td>\n",
       "      <td>6</td>\n",
       "      <td>305</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>rs150058293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BACH2</td>\n",
       "      <td>ENSG00000112182</td>\n",
       "      <td>6</td>\n",
       "      <td>1546</td>\n",
       "      <td>0.998960</td>\n",
       "      <td>rs72928038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H2BC15</td>\n",
       "      <td>ENSG00000233822</td>\n",
       "      <td>6</td>\n",
       "      <td>658</td>\n",
       "      <td>0.998254</td>\n",
       "      <td>rs77619023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HLA-DPA1</td>\n",
       "      <td>ENSG00000231389</td>\n",
       "      <td>6</td>\n",
       "      <td>418</td>\n",
       "      <td>0.996692</td>\n",
       "      <td>rs116108798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RAB5B</td>\n",
       "      <td>ENSG00000111540</td>\n",
       "      <td>12</td>\n",
       "      <td>346</td>\n",
       "      <td>0.993501</td>\n",
       "      <td>rs1873914</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name          gene_id  chromosome  tested_snps      pph4      top_snp\n",
       "0    TRIM26  ENSG00000234127           6          305  1.000000  rs150058293\n",
       "1     BACH2  ENSG00000112182           6         1546  0.998960   rs72928038\n",
       "2    H2BC15  ENSG00000233822           6          658  0.998254   rs77619023\n",
       "3  HLA-DPA1  ENSG00000231389           6          418  0.996692  rs116108798\n",
       "4     RAB5B  ENSG00000111540          12          346  0.993501    rs1873914"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_reuma.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>tested_snps</th>\n",
       "      <th>pph4</th>\n",
       "      <th>top_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SHE</td>\n",
       "      <td>ENSG00000169291</td>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>rs2229238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IL2RA</td>\n",
       "      <td>ENSG00000134460</td>\n",
       "      <td>10</td>\n",
       "      <td>1015</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>rs61839660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HDLBP</td>\n",
       "      <td>ENSG00000115677</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>rs7570017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPRYD3</td>\n",
       "      <td>ENSG00000167778</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>rs7313065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IL6R</td>\n",
       "      <td>ENSG00000160712</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>0.999996</td>\n",
       "      <td>rs2229238</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name          gene_id  chromosome  tested_snps      pph4     top_snp\n",
       "0       SHE  ENSG00000169291           1           29  1.000000   rs2229238\n",
       "1     IL2RA  ENSG00000134460          10         1015  1.000000  rs61839660\n",
       "2     HDLBP  ENSG00000115677           2            6  0.999999   rs7570017\n",
       "3    SPRYD3  ENSG00000167778          12            3  0.999997   rs7313065\n",
       "4      IL6R  ENSG00000160712           1           30  0.999996   rs2229238"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_t1d.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out genes where the tested snps were lower than 100\n",
    "df_t1d = df_t1d[df_t1d['tested_snps'] > 100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RA colocalization has 0 missing values\n",
      "T1D colocalization has 0 missing values\n"
     ]
    }
   ],
   "source": [
    "# Check for missing values\n",
    "print(f'RA colocalization has {df_reuma.isnull().sum().sum()} missing values')\n",
    "print(f'T1D colocalization has {df_t1d.isnull().sum().sum()} missing values')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6     10\n",
       "16     6\n",
       "17     6\n",
       "19     5\n",
       "7      2\n",
       "10     2\n",
       "2      1\n",
       "4      1\n",
       "8      1\n",
       "12     1\n",
       "13     1\n",
       "Name: chromosome, dtype: int64"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Amount of genes per chromosome for type 1 diabetes\n",
    "df_t1d['chromosome'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6     11\n",
       "2      3\n",
       "8      3\n",
       "12     3\n",
       "17     3\n",
       "22     3\n",
       "1      2\n",
       "5      2\n",
       "3      1\n",
       "9      1\n",
       "11     1\n",
       "15     1\n",
       "19     1\n",
       "Name: chromosome, dtype: int64"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Amount of genes per chromosome for rheumatoid arthritis\n",
    "df_reuma['chromosome'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amount of colocalizing genes found for T1D is 36\n",
      "Amount of colocalizing genes found for RA is 35\n"
     ]
    }
   ],
   "source": [
    "print(f'Amount of colocalizing genes found for T1D is {len(df_t1d)}')\n",
    "print(f'Amount of colocalizing genes found for RA is {len(df_reuma)}')\n",
    "#rs72928038"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   gene_name          gene_id  chromosome  tested_snps      pph4     top_snp\n",
      "7      BACH2  ENSG00000112182           6          753  0.999858  rs72928038\n",
      "22     ERBB2  ENSG00000141736          17          508  0.974361   rs2941522\n",
      "26      SUOX  ENSG00000139531          12          246  0.967435  rs34415530\n",
      "36     CTLA4  ENSG00000163599           2          587  0.927506   rs3087243\n",
      "47    ORMDL3  ENSG00000172057          17          632  0.855811  rs56380902\n",
      "48     GSDMB  ENSG00000073605          17          639  0.852767  rs56380902\n",
      "\n",
      "The amount of genes that overlap between the two analyses are: 6\n"
     ]
    }
   ],
   "source": [
    "# Overlapping genes between both colocalization analyses\n",
    "overlap_gene = df_t1d[df_t1d['gene_id'].isin(df_reuma['gene_id'])]\n",
    "print(f'{overlap_gene}\\n')\n",
    "print(f'The amount of genes that overlap between the two analyses are: {len(overlap_gene)}')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BACH2 and CTLA4 share the same top variant in both RA and T1D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>tested_snps</th>\n",
       "      <th>pph4</th>\n",
       "      <th>top_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BACH2</td>\n",
       "      <td>ENSG00000112182</td>\n",
       "      <td>6</td>\n",
       "      <td>753</td>\n",
       "      <td>0.999858</td>\n",
       "      <td>rs72928038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>CTLA4</td>\n",
       "      <td>ENSG00000163599</td>\n",
       "      <td>2</td>\n",
       "      <td>587</td>\n",
       "      <td>0.927506</td>\n",
       "      <td>rs3087243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name          gene_id  chromosome  tested_snps      pph4     top_snp\n",
       "7      BACH2  ENSG00000112182           6          753  0.999858  rs72928038\n",
       "36     CTLA4  ENSG00000163599           2          587  0.927506   rs3087243"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_t1d[df_t1d['top_snp'].isin(['rs72928038', 'rs3087243'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>tested_snps</th>\n",
       "      <th>pph4</th>\n",
       "      <th>top_snp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BACH2</td>\n",
       "      <td>ENSG00000112182</td>\n",
       "      <td>6</td>\n",
       "      <td>1546</td>\n",
       "      <td>0.998960</td>\n",
       "      <td>rs72928038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CTLA4</td>\n",
       "      <td>ENSG00000163599</td>\n",
       "      <td>2</td>\n",
       "      <td>665</td>\n",
       "      <td>0.932209</td>\n",
       "      <td>rs3087243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name          gene_id  chromosome  tested_snps      pph4     top_snp\n",
       "1      BACH2  ENSG00000112182           6         1546  0.998960  rs72928038\n",
       "19     CTLA4  ENSG00000163599           2          665  0.932209   rs3087243"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_reuma[df_reuma['top_snp'].isin(['rs72928038', 'rs3087243'])]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_barplot(df, name):   \n",
    "    genes = df.gene_name\n",
    "    pph4 = df.pph4\n",
    "    p = figure(x_range=genes, height=400, plot_width=800,\n",
    "               title=f'PP4 values of genes of {name} colocalization results',\n",
    "               toolbar_location=None)\n",
    "    p.vbar(x=genes, top=pph4, width=0.85)\n",
    "    p.yaxis.axis_label = \"PPH4\"\n",
    "    p.xaxis.axis_label = \"Genes\"\n",
    "    p.xgrid.grid_line_color=None\n",
    "    p.xaxis.major_label_orientation = 1.2\n",
    "\n",
    "    p.y_range.start = 0.8\n",
    "    show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"67d133c6-b3bd-46d9-b92f-66228c6c6b55\" data-root-id=\"2059\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function embed_document(root) {\n    \n  var docs_json = {\"03bd0afc-2eab-4ff2-abf2-fd03e070bba6\":{\"defs\":[],\"roots\":{\"references\":[{\"attributes\":{\"below\":[{\"id\":\"2070\"}],\"center\":[{\"id\":\"2072\"},{\"id\":\"2076\"}],\"height\":400,\"left\":[{\"id\":\"2073\"}],\"renderers\":[{\"id\":\"2094\"}],\"title\":{\"id\":\"2060\"},\"toolbar\":{\"id\":\"2084\"},\"toolbar_location\":null,\"width\":800,\"x_range\":{\"id\":\"2062\"},\"x_scale\":{\"id\":\"2066\"},\"y_range\":{\"id\":\"2064\"},\"y_scale\":{\"id\":\"2068\"}},\"id\":\"2059\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{\"data\":{\"top\":{\"__ndarray__\":\"G55X2P//7z+hz6h11v7vP4S9Ci73+O8/7NSCm8n47z/0LCHxydDvP0MbOWCaru8/VsMDAPgt7z/20t6HcxjvP3g5Y/OqCO8/YOta+zn17j+G83kB/ufuPy4/AyxT5u4/T6dos++K7j9mgvuzxInuP8oY5c00Z+4/w1ezCjAk7j8PpWzxafztPzyzYWDDte0/KKAbYSGu7T8MWf0YpKXtP4OAFsnWe+0/fvvLqOtK7T+X+vlP+RPtP4Sw+hX5sOw/QRFE18yc7D+DNdtitW/sP8DxyXOHROw/jA8Cs1HX6z/IcV7b4NbrP73GR7zMYus/zVkahN5J6z+jUKrJPSvrPzmw1S1EC+s/B3vs9J2G6j9II/TBP4bqP5vKTev/Deo/\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[36]},\"x\":[\"IL2RA\",\"BACH2\",\"HLA-H\",\"ENSG00000230521\",\"NTN5\",\"ANKRD16\",\"ERBB2\",\"LY6G6C\",\"UBAC2-AS1\",\"SUOX\",\"CTSB\",\"ENSG00000269487\",\"IKZF1\",\"HLA-U\",\"IKZF3\",\"SH2B1\",\"HLA-DQA2\",\"PRKD2\",\"CTLA4\",\"CENPW\",\"DGKQ\",\"HSPA1A\",\"ENSG00000264058\",\"TUFM\",\"TYK2\",\"SMARCE1\",\"SGF29\",\"HOTAIRM1\",\"HLA-DQB1\",\"ORMDL3\",\"GSDMB\",\"ENSG00000261476\",\"ATP2A1\",\"NUPR1\",\"IER3\",\"KNOWN\"]},\"selected\":{\"id\":\"2183\"},\"selection_policy\":{\"id\":\"2184\"}},\"id\":\"2091\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"start\":0.8},\"id\":\"2064\",\"type\":\"DataRange1d\"},{\"attributes\":{},\"id\":\"2080\",\"type\":\"SaveTool\"},{\"attributes\":{},\"id\":\"2068\",\"type\":\"LinearScale\"},{\"attributes\":{\"axis_label\":\"Genes\",\"formatter\":{\"id\":\"2177\"},\"major_label_orientation\":1.2,\"major_label_policy\":{\"id\":\"2178\"},\"ticker\":{\"id\":\"2071\"}},\"id\":\"2070\",\"type\":\"CategoricalAxis\"},{\"attributes\":{\"factors\":[\"IL2RA\",\"BACH2\",\"HLA-H\",\"ENSG00000230521\",\"NTN5\",\"ANKRD16\",\"ERBB2\",\"LY6G6C\",\"UBAC2-AS1\",\"SUOX\",\"CTSB\",\"ENSG00000269487\",\"IKZF1\",\"HLA-U\",\"IKZF3\",\"SH2B1\",\"HLA-DQA2\",\"PRKD2\",\"CTLA4\",\"CENPW\",\"DGKQ\",\"HSPA1A\",\"ENSG00000264058\",\"TUFM\",\"TYK2\",\"SMARCE1\",\"SGF29\",\"HOTAIRM1\",\"HLA-DQB1\",\"ORMDL3\",\"GSDMB\",\"ENSG00000261476\",\"ATP2A1\",\"NUPR1\",\"IER3\",\"KNOWN\"]},\"id\":\"2062\",\"type\":\"FactorRange\"},{\"attributes\":{},\"id\":\"2066\",\"type\":\"CategoricalScale\"},{\"attributes\":{\"axis_label\":\"PPH4\",\"formatter\":{\"id\":\"2180\"},\"major_label_policy\":{\"id\":\"2181\"},\"ticker\":{\"id\":\"2074\"}},\"id\":\"2073\",\"type\":\"LinearAxis\"},{\"attributes\":{},\"id\":\"2082\",\"type\":\"HelpTool\"},{\"attributes\":{},\"id\":\"2071\",\"type\":\"CategoricalTicker\"},{\"attributes\":{\"axis\":{\"id\":\"2070\"},\"grid_line_color\":null,\"ticker\":null},\"id\":\"2072\",\"type\":\"Grid\"},{\"attributes\":{\"axis\":{\"id\":\"2073\"},\"dimension\":1,\"ticker\":null},\"id\":\"2076\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"2074\",\"type\":\"BasicTicker\"},{\"attributes\":{\"active_multi\":null,\"tools\":[{\"id\":\"2077\"},{\"id\":\"2078\"},{\"id\":\"2079\"},{\"id\":\"2080\"},{\"id\":\"2081\"},{\"id\":\"2082\"}]},\"id\":\"2084\",\"type\":\"Toolbar\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"top\":{\"field\":\"top\"},\"width\":{\"value\":0.85},\"x\":{\"field\":\"x\"}},\"id\":\"2093\",\"type\":\"VBar\"},{\"attributes\":{},\"id\":\"2078\",\"type\":\"WheelZoomTool\"},{\"attributes\":{},\"id\":\"2077\",\"type\":\"PanTool\"},{\"attributes\":{\"overlay\":{\"id\":\"2083\"}},\"id\":\"2079\",\"type\":\"BoxZoomTool\"},{\"attributes\":{},\"id\":\"2081\",\"type\":\"ResetTool\"},{\"attributes\":{},\"id\":\"2177\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{\"source\":{\"id\":\"2091\"}},\"id\":\"2095\",\"type\":\"CDSView\"},{\"attributes\":{},\"id\":\"2178\",\"type\":\"AllLabels\"},{\"attributes\":{\"data_source\":{\"id\":\"2091\"},\"glyph\":{\"id\":\"2092\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"2093\"},\"view\":{\"id\":\"2095\"}},\"id\":\"2094\",\"type\":\"GlyphRenderer\"},{\"attributes\":{},\"id\":\"2180\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{\"fill_color\":{\"value\":\"#1f77b4\"},\"line_color\":{\"value\":\"#1f77b4\"},\"top\":{\"field\":\"top\"},\"width\":{\"value\":0.85},\"x\":{\"field\":\"x\"}},\"id\":\"2092\",\"type\":\"VBar\"},{\"attributes\":{},\"id\":\"2184\",\"type\":\"UnionRenderers\"},{\"attributes\":{},\"id\":\"2181\",\"type\":\"AllLabels\"},{\"attributes\":{\"text\":\"PP4 values of genes of T1D colocalization results\"},\"id\":\"2060\",\"type\":\"Title\"},{\"attributes\":{},\"id\":\"2183\",\"type\":\"Selection\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":0.5,\"fill_color\":\"lightgrey\",\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":1.0,\"line_color\":\"black\",\"line_dash\":[4,4],\"line_width\":2,\"right_units\":\"screen\",\"syncable\":false,\"top_units\":\"screen\"},\"id\":\"2083\",\"type\":\"BoxAnnotation\"}],\"root_ids\":[\"2059\"]},\"title\":\"Bokeh Application\",\"version\":\"2.3.2\"}};\n  var render_items = [{\"docid\":\"03bd0afc-2eab-4ff2-abf2-fd03e070bba6\",\"root_ids\":[\"2059\"],\"roots\":{\"2059\":\"67d133c6-b3bd-46d9-b92f-66228c6c6b55\"}}];\n  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n\n  }\n  if (root.Bokeh !== undefined) {\n    embed_document(root);\n  } else {\n    var attempts = 0;\n    var timer = setInterval(function(root) {\n      if (root.Bokeh !== undefined) {\n        clearInterval(timer);\n        embed_document(root);\n      } else {\n        attempts++;\n        if (attempts > 100) {\n          clearInterval(timer);\n          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n        }\n      }\n    }, 10, root)\n  }\n})(window);",
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "2059"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "make_barplot(df_t1d, 'T1D')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "  <div class=\"bk-root\" id=\"dffee01e-c525-4b23-b1b0-d108dce6a0bf\" data-root-id=\"2423\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function embed_document(root) {\n    \n  var docs_json = {\"19a3eff6-12b1-428a-99ac-dabc8659ccdb\":{\"defs\":[],\"roots\":{\"references\":[{\"attributes\":{\"below\":[{\"id\":\"2434\"}],\"center\":[{\"id\":\"2436\"},{\"id\":\"2440\"}],\"height\":400,\"left\":[{\"id\":\"2437\"}],\"renderers\":[{\"id\":\"2458\"}],\"title\":{\"id\":\"2424\"},\"toolbar\":{\"id\":\"2448\"},\"toolbar_location\":null,\"width\":800,\"x_range\":{\"id\":\"2426\"},\"x_scale\":{\"id\":\"2430\"},\"y_range\":{\"id\":\"2428\"},\"y_scale\":{\"id\":\"2432\"}},\"id\":\"2423\",\"subtype\":\"Figure\",\"type\":\"Plot\"},{\"attributes\":{},\"id\":\"2568\",\"type\":\"UnionRenderers\"},{\"attributes\":{\"active_multi\":null,\"tools\":[{\"id\":\"2441\"},{\"id\":\"2442\"},{\"id\":\"2443\"},{\"id\":\"2444\"},{\"id\":\"2445\"},{\"id\":\"2446\"}]},\"id\":\"2448\",\"type\":\"Toolbar\"},{\"attributes\":{\"start\":0.8},\"id\":\"2428\",\"type\":\"DataRange1d\"},{\"attributes\":{\"axis_label\":\"Genes\",\"formatter\":{\"id\":\"2561\"},\"major_label_orientation\":1.2,\"major_label_policy\":{\"id\":\"2562\"},\"ticker\":{\"id\":\"2435\"}},\"id\":\"2434\",\"type\":\"CategoricalAxis\"},{\"attributes\":{},\"id\":\"2444\",\"type\":\"SaveTool\"},{\"attributes\":{\"data\":{\"top\":{\"__ndarray__\":\"AAAAAAAA8D9Uaj+5e/fvP3tv0mCy8e8/rd2W0+Xk7z9/oVQ1w8rvP8xJeyCAvu8/hyL98im87z8SjEbS/rbvP4pJMpG4qO8/0K1caW177z9RORPFfW7vP/b09STJSe8/v/Q2Z/X57j8hifRKGL7uPykN5m+joe4/ejGo14RR7j8nRkDNUDruP2YOkonqLu4/n78gblwT7j9ygmL4ptTtP8N13RmDq+0/BQlDpd2d7T927T4MBGPtP6I12gklX+0/Jpp8RPZN7T8KgNZDLPDsPx4Pj6fq6ew/Lyfu+uPk7D9lVZl2znrsP5S6Bl44/us/KBGWy4Ix6z+DOfcFKNTqP5/6O0ihsuo/Ujecp32l6j/r4juR4J7qPw==\",\"dtype\":\"float64\",\"order\":\"little\",\"shape\":[35]},\"x\":[\"TRIM26\",\"BACH2\",\"H2BC15\",\"HLA-DPA1\",\"RAB5B\",\"LINC00824\",\"ANKRD55\",\"CCR6\",\"KEAP1\",\"ENSG00000267520\",\"PSTPIP1\",\"ENSG00000269468\",\"HLA-F\",\"EOMES\",\"SUOX\",\"MMEL1\",\"AFF3\",\"ERBB2\",\"BLK\",\"CTLA4\",\"DAP\",\"PHF19\",\"UQCC2\",\"SYNGR1\",\"CXCR5\",\"PPP1R18\",\"UBE2L3\",\"ENSG00000199332\",\"CLIC1\",\"OS9\",\"CCDC116\",\"FAM167A\",\"ORMDL3\",\"DHX16\",\"GSDMB\"]},\"selected\":{\"id\":\"2567\"},\"selection_policy\":{\"id\":\"2568\"}},\"id\":\"2455\",\"type\":\"ColumnDataSource\"},{\"attributes\":{\"factors\":[\"TRIM26\",\"BACH2\",\"H2BC15\",\"HLA-DPA1\",\"RAB5B\",\"LINC00824\",\"ANKRD55\",\"CCR6\",\"KEAP1\",\"ENSG00000267520\",\"PSTPIP1\",\"ENSG00000269468\",\"HLA-F\",\"EOMES\",\"SUOX\",\"MMEL1\",\"AFF3\",\"ERBB2\",\"BLK\",\"CTLA4\",\"DAP\",\"PHF19\",\"UQCC2\",\"SYNGR1\",\"CXCR5\",\"PPP1R18\",\"UBE2L3\",\"ENSG00000199332\",\"CLIC1\",\"OS9\",\"CCDC116\",\"FAM167A\",\"ORMDL3\",\"DHX16\",\"GSDMB\"]},\"id\":\"2426\",\"type\":\"FactorRange\"},{\"attributes\":{},\"id\":\"2432\",\"type\":\"LinearScale\"},{\"attributes\":{\"axis_label\":\"PPH4\",\"formatter\":{\"id\":\"2564\"},\"major_label_policy\":{\"id\":\"2565\"},\"ticker\":{\"id\":\"2438\"}},\"id\":\"2437\",\"type\":\"LinearAxis\"},{\"attributes\":{},\"id\":\"2430\",\"type\":\"CategoricalScale\"},{\"attributes\":{},\"id\":\"2446\",\"type\":\"HelpTool\"},{\"attributes\":{},\"id\":\"2435\",\"type\":\"CategoricalTicker\"},{\"attributes\":{\"axis\":{\"id\":\"2434\"},\"grid_line_color\":null,\"ticker\":null},\"id\":\"2436\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"2564\",\"type\":\"BasicTickFormatter\"},{\"attributes\":{\"axis\":{\"id\":\"2437\"},\"dimension\":1,\"ticker\":null},\"id\":\"2440\",\"type\":\"Grid\"},{\"attributes\":{},\"id\":\"2438\",\"type\":\"BasicTicker\"},{\"attributes\":{},\"id\":\"2442\",\"type\":\"WheelZoomTool\"},{\"attributes\":{},\"id\":\"2441\",\"type\":\"PanTool\"},{\"attributes\":{\"overlay\":{\"id\":\"2447\"}},\"id\":\"2443\",\"type\":\"BoxZoomTool\"},{\"attributes\":{},\"id\":\"2445\",\"type\":\"ResetTool\"},{\"attributes\":{},\"id\":\"2561\",\"type\":\"CategoricalTickFormatter\"},{\"attributes\":{\"fill_color\":{\"value\":\"#1f77b4\"},\"line_color\":{\"value\":\"#1f77b4\"},\"top\":{\"field\":\"top\"},\"width\":{\"value\":0.85},\"x\":{\"field\":\"x\"}},\"id\":\"2456\",\"type\":\"VBar\"},{\"attributes\":{},\"id\":\"2567\",\"type\":\"Selection\"},{\"attributes\":{},\"id\":\"2562\",\"type\":\"AllLabels\"},{\"attributes\":{\"source\":{\"id\":\"2455\"}},\"id\":\"2459\",\"type\":\"CDSView\"},{\"attributes\":{\"fill_alpha\":{\"value\":0.1},\"fill_color\":{\"value\":\"#1f77b4\"},\"line_alpha\":{\"value\":0.1},\"line_color\":{\"value\":\"#1f77b4\"},\"top\":{\"field\":\"top\"},\"width\":{\"value\":0.85},\"x\":{\"field\":\"x\"}},\"id\":\"2457\",\"type\":\"VBar\"},{\"attributes\":{\"data_source\":{\"id\":\"2455\"},\"glyph\":{\"id\":\"2456\"},\"hover_glyph\":null,\"muted_glyph\":null,\"nonselection_glyph\":{\"id\":\"2457\"},\"view\":{\"id\":\"2459\"}},\"id\":\"2458\",\"type\":\"GlyphRenderer\"},{\"attributes\":{\"text\":\"PP4 values of genes of RA colocalization results\"},\"id\":\"2424\",\"type\":\"Title\"},{\"attributes\":{},\"id\":\"2565\",\"type\":\"AllLabels\"},{\"attributes\":{\"bottom_units\":\"screen\",\"fill_alpha\":0.5,\"fill_color\":\"lightgrey\",\"left_units\":\"screen\",\"level\":\"overlay\",\"line_alpha\":1.0,\"line_color\":\"black\",\"line_dash\":[4,4],\"line_width\":2,\"right_units\":\"screen\",\"syncable\":false,\"top_units\":\"screen\"},\"id\":\"2447\",\"type\":\"BoxAnnotation\"}],\"root_ids\":[\"2423\"]},\"title\":\"Bokeh Application\",\"version\":\"2.3.2\"}};\n  var render_items = [{\"docid\":\"19a3eff6-12b1-428a-99ac-dabc8659ccdb\",\"root_ids\":[\"2423\"],\"roots\":{\"2423\":\"dffee01e-c525-4b23-b1b0-d108dce6a0bf\"}}];\n  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n\n  }\n  if (root.Bokeh !== undefined) {\n    embed_document(root);\n  } else {\n    var attempts = 0;\n    var timer = setInterval(function(root) {\n      if (root.Bokeh !== undefined) {\n        clearInterval(timer);\n        embed_document(root);\n      } else {\n        attempts++;\n        if (attempts > 100) {\n          clearInterval(timer);\n          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n        }\n      }\n    }, 10, root)\n  }\n})(window);",
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "2423"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "make_barplot(df_reuma, 'RA')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
